### PR TITLE
refactoring flybase ingest

### PR DESCRIPTION
**Pull request by kshefchek**
This PR is a large overhaul of the original flybase ingest, simplifying the model to include the following associations:
- allele-gene
- allele model_of disease
- allele-phenotype associations

This removes a number of things, but of interest
- Fly genotypes, which are stored in the relational db but are not available on the flybase app
- Blank node "models" are being replaced by alleles
- Stocks, stock-genotype, [environments](https://github.com/monarch-initiative/dipper/blob/master/dipper/sources/FlyBase.py#L542)

Some integration notes:
Allele-gene and allele-phenotype can be joined to create gene-phenotype associations, with the caveat that we don't want make this inference for transgenes.

### CODE REVIEW

**Comment by dosumis** [comment]
Wondering why we have stock info?  I  guess this is proof that organisms are available for experiments, which could  be useful.   But do we take advantage of this?

> Some integration notes:
> Allele-gene and allele-phenotype can be joined to create gene-phenotype associations, with the caveat that we don't want make this inference for transgenes.

Worth looking into this further.  There might be enough info in the files to some crude filtering before joining.  First thing would be to eliminate foreign genes.  Might also be worth trying to eliminate transgenes of Drosophila genes, although some phenotypes from  these might be informative about normal function.

**Comment by kshefchek** [comment]
@dosumis do you think we can rely on the gene label nomenclature? non melanogaster species seem to have prefixes as defined in the file here ftp://ftp.flybase.net/releases/current/precomputed_files/species/species.ab.gz

**Comment by kshefchek** [comment]
I still need to add some tests, but the ingest code is ready for review.  The ingest now takes ~25 minutes versus 5 hours previously, so it's feasible to run it locally and investigate the output (144M turtle file).

File: dipper/sources/PostgreSQLSource.py [hunk 1]
```diff
@@ -69,61 +68,53 @@ def fetch_from_pgdb(self, tables, cxn, limit=None, force=False):
...
+                if (filerowcount - 1) < tablerowcount:
+                    raise Exception(
+                        "Download from %s failed, %s != %s",
+                        cxn['host'] + ':' + cxn['database'],
+                        (filerowcount - 1),
+                        tablerowcount
+                    )
+                elif (filerowcount - 1) > tablerowcount:
+                    LOG.warning(
+                        "Fetched from %s more rows in file (%s) than reported in count(%s)",
```
Comment (justaddcoffee) [inline dipper/sources/PostgreSQLSource.py]: line's a bit too long (per our endless pylint convo haha)

Comment (kshefchek) [reply dipper/sources/PostgreSQLSource.py]: thanks! forgot to lint that file

File: dipper/sources/PostgreSQLSource.py [hunk 2]
```diff
@@ -135,20 +131,20 @@ def fetch_query_from_pgdb(self, qname, query, con, cxn, limit=None, force=False)
...
-                LOG.warning(
-                    "Fetched from %s more rows in file (%s) than reported in count(%s)",
-                    cxn['host'] + ':'+cxn['database'], (filerowcount-1), tablerowcount)
-
-    # TODO generalize this to a set of utils
+                "Download from {} failed, {} != {}"
+                .format(cxn['host'] + ':' + cxn['database'],
+                        (filerowcount-1), tablerowcount))
+        if (filerowcount-1) > tablerowcount:
+            LOG.warning(
```
Comment (justaddcoffee) [inline dipper/sources/PostgreSQLSource.py]: Getting an indentation error here: 

PostgreSQLSource.py:144:0: C0330: Wrong hanging indentation (remove 4 spaces).
                    "Download from %s failed, %s != %s", cxn['host'] + ':' +
                |   ^ (bad-continuation)
PostgreSQLSource.py:145:0: C0330: Wrong hanging indentation (remove 4 spaces).
                    cxn['database'], (filerowcount-1), tablerowcount)
                |   ^ (bad-continuation)

Comment (kshefchek) [reply dipper/sources/PostgreSQLSource.py]: odd, I'm not getting this warning, are you running it on 44a9c5b?

Comment (justaddcoffee) [reply dipper/sources/PostgreSQLSource.py]: Yep, sorry, needed to do a git pull, I was on [this previous commit](https://github.com/monarch-initiative/dipper/pull/760/commits/0e6f91e25c1a794c2f761257e4d2f99726404512)

File: dipper/sources/FlyBase.py [hunk 3]
```diff
@@ -20,180 +19,110 @@
...
-        """
     }
 
-    FLYFTP = 'ftp://ftp.flybase.net/releases/current/precomputed_files/human_disease/'
     files = {
-        'disease_models': {
+        'disease_model': {
             'file': 'allele_human_disease_model_data.tsv.gz',
-            'url':  FLYFTP + 'allele_human_disease_model_data_fb_*.tsv.gz'
+            'url':  r'human_disease/allele_human_disease_model_data_fb_.*\.tsv\.gz',
```
Comment (justaddcoffee) [inline dipper/sources/FlyBase.py]: change to: 
human_disease/disease_model_annotations_fb.*.tsv.gz

File: dipper/sources/FlyBase.py [hunk 4]
```diff
@@ -20,180 +19,110 @@
...
+    from which we process genotype and phenotype data about the fruit fly.
+
+    Here, we connect to their public database and download preprocessed files
+
+    Queries from the relational db
+    1. allele-phenotype data: ../../sources/sql/fb/allele_phenotype.sql
+    2. gene dbxrefs: ../../resources/sql/fb/gene_xref.sql
+
+    Downloads:
+    1. allele_human_disease_model_data_fb_*.tsv.gz - models of disease
```
Comment (justaddcoffee) [inline dipper/sources/FlyBase.py]: change to:
    1. disease_model_annotations_fb_*.tsv.gz - models of disease

File: dipper/sources/FlyBase.py [hunk 5]
```diff
@@ -268,1844 +152,522 @@ def fetch(self, is_dl_forced=False):
...
-        with open(raw, 'r') as f:
-            f.readline()  # read the header row; skip
-            filereader = csv.reader(f, delimiter='\t', quotechar='\"')
-            for line in filereader:
-                (feature_relationship_id, subject_id, object_id, name, rank,
-                 value) = line
-
-                if name == 'derived_tp_assoc_alleles':
-                    # derived_tp_assoc_alleles
-                    self.f
```
Comment (justaddcoffee) [inline dipper/sources/FlyBase.py]: I'm getting an assertion error here, I think to do with missing columns
```
ERROR:dipper.sources.Source:
Expected header: ['FBal_ID', 'AlleleSymbol', 'DOID_qualifier', 'DOID_term', 'DOID_ID', 'Evidence/interacting_alleles', 'Reference_FBid']
Recieved header: [' FBgn ID', 'Gene symbol', 'HGNC ID', 'DO qualifier', 'DO ID', 'DO term', 'Allele used in model (FBal ID)', 'Allele used in model (symbol)', 'Based on orthology with (HGNC ID)', 'Based on orthology with (symbol)', 'Evidence/interacting alleles', 'Reference (FBrf ID)']
ERROR:dipper.sources.Source:Missing: {'FBal_ID', 'DOID_qualifier', 'DOID_ID', 'Evidence/interacting_alleles', 'AlleleSymbol', 'DOID_term', 'Reference_FBid'}
Traceback (most recent call last):
  File "./dipper-etl.py", line 280, in <module>
    main()
  File "./dipper-etl.py", line 240, in main
    mysource.parse(args.limit)
  File "/Users/jtr4v/projects/MI/dipper/dipper-main/dipper/sources/FlyBase.py", line 191, in parse
    self._process_disease_model(limit)
  File "/Users/jtr4v/projects/MI/dipper/dipper-main/dipper/sources/FlyBase.py", line 602, in _process_disease_model
    self.check_fileheader(col, row)
  File "/Users/jtr4v/projects/MI/dipper/dipper-main/dipper/sources/Source.py", line 839, in check_fileheader
    raise AssertionError('Incomming headers are missing expected column.')
AssertionError: Incomming headers are missing expected column.
```

**Comment by kshefchek** [comment]
Looks like flybase had a new release yesterday and has adjusted some of the precomputed files, will update the code accordingly

**Comment by justaddcoffee** [comment]
Seems to work like a charm now on my end

File: dipper/sources/FlyBase.py [hunk 6]
```diff
@@ -268,1844 +157,526 @@ def fetch(self, is_dl_forced=False):
...
-        These are all of the genomic features genes, variations,
-        transgenes, etc
-        :param limit:
-        :return:
+                allele_curie = 'FlyBase:' + allele_id
 
-        """
+                # a alphanumeric id followed by a colon then
+                # any character but a colon bordered by @s
+                term_regex = re.compile(r'@([\w]*):[^:@]*@')
```
Comment (TomConlin) [inline dipper/sources/FlyBase.py]: recommend taking the regex compiles out of the loop

File: dipper/sources/FlyBase.py [hunk 7]
```diff
@@ -268,1844 +157,526 @@ def fetch(self, is_dl_forced=False):
...
-                        did = dbxrefs.get(d)
-                        if did.endswith('&class=protein'):
-                            did = did[0:len(dbxrefs)-15]
-                        # don't make something sameAs itself
-                        if did == feature_id:
+                if len(allele_prefix) == 1:
+                    try:
+                        if species_map[allele_prefix[0]][0] == 'drosophilid':
+                            geno.addAllele(allele_curie, allele_label)
+                            geno.addTaxon(species_map[allele_prefix[0]][1], allele_curie)
```
Comment (TomConlin) [inline dipper/sources/FlyBase.py]: 89 char, no big deal

File: dipper/sources/Source.py [hunk 8]
```diff
@@ -322,18 +323,20 @@ def checkIfRemoteIsNewer(self, remote, local, headers):
...
                 if fstat[ST_SIZE] < size:
                     LOG.info("New Remote File exists")
-                    return True
+                    is_remote_newer = True
                 if fstat[ST_SIZE] > size:
                     LOG.warning("New Remote File exists but it is SMALLER")
-                    return True
+                    is_remote_newer = True
                 # filesize is a fairly imperfect metric here
                 LOG.info("New Remote fFle has same filesize--will not download")
```
Comment (TomConlin) [inline dipper/sources/Source.py]: fFle -> file   (most likely my typo)

**Comment by TomConlin** [comment]
+1  left a few minor notes inline

**Comment by justaddcoffee** [comment]
+1

**Comment by kshefchek** [comment]
Still working out what IDs to test for various edge cases, but in the meantime there a few things to be resolved:

1. How do we handle "phenotype manifest in" data?
- Link them to the anatomical entity via 'located in' or 'expressed in' and/or;
- Link them to UPheno classes with 'has phenotype', eg https://monarchinitiative.org/phenotype/FBbt:00000004PHENOTYPE or;
- Ignore all together

2. How do we model transgenic alleles, for example http://flybase.org/reports/FBal0256668
- Even though this allele is annotated to humans, should we annotate it to _D. melanogaster_?

**Comment by kshefchek** [comment]
I've added some tests and sample data sets for a couple of alleles.  Proposing that we get this merged and then create a ticket and circle back on the anatomy/phenotype data.

**Comment by dosumis** [comment]
> How do we model transgenic alleles, for example http://flybase.org/reports/FBal0256668
> Even though this allele is annotated to human, should we annotate it to D. melanogaster?

No!  Probably best to just ditch all non-Dmel transgenes.  There's also a pretty strong case for ditching transgenes altogether - with the exception of RNAi - as most will be gain of function and many expressed in places they are not normally found.    My suggestion:  First pass - ignore all transgenes.  Make a ticket for flagging RNAi transgenes & I'll find a way to extract this from public postgress (-> CSV).

BTW - how are you flagging transgenes in the first place?
